cmd/kubeapps-apis/server/server.go

```diff
@@ -88,3 +109,88 @@ type gwHandlerArgs struct {
 	addr        string
 	dialOptions []grpc.DialOption
 }
+
+// registerPlugins opens each plugin, looks up the register function and calls it with the registrar.
+func registerPlugins(pluginPaths []string, grpcReg grpc.ServiceRegistrar, gwArgs gwHandlerArgs) error {
+	for _, pluginPath := range pluginPaths {
+		p, err := plugin.Open(pluginPath)
+		if err != nil {
+			return fmt.Errorf("unable to open plugin %q: %w", pluginPath, err)
+		}
+
+		if err = registerGRPC(p, pluginPath, grpcReg); err != nil {
+			return err
+		}
+
+		if err = registerHTTP(p, pluginPath, gwArgs); err != nil {
+			return err
+		}
+
+		log.Infof("Successfully registered plugin %q", pluginPath)
+	}
+	return nil
+}
+
+// registerGRPC finds and calls the required function for registering the plugin for the GRPC server.
+func registerGRPC(p *plugin.Plugin, pluginPath string, registrar grpc.ServiceRegistrar) error {
+	grpcRegFn, err := p.Lookup(grpcRegisterFunction)
+	if err != nil {
+		return fmt.Errorf("unable to lookup %q for %q: %w", grpcRegisterFunction, pluginPath, err)
+	}
+	type grpcRegisterFunctionType = func(grpc.ServiceRegistrar)
+	grpcFn, ok := grpcRegFn.(grpcRegisterFunctionType)
+	if !ok {
+		var dummyFn grpcRegisterFunctionType = func(grpc.ServiceRegistrar) {}
+		return fmt.Errorf("unable to use %q in plugin %q due to mismatched signature.\nwant: %T\ngot: %T", grpcRegisterFunction, pluginPath, dummyFn, grpcRegFn)
+	}
+	grpcFn(registrar)
+	return nil
+}
+
+// registerHTTP finds and calls the required function for registering the plugin for the HTTP gateway server.
+func registerHTTP(p *plugin.Plugin, pluginPath string, gwArgs gwHandlerArgs) error {
+	gwRegFn, err := p.Lookup(gatewayRegisterFunction)
+	if err != nil {
+		return fmt.Errorf("unable to lookup %q for %q: %w", gatewayRegisterFunction, pluginPath, err)
+	}
+	type gatewayRegisterFunctionType = func(context.Context, *runtime.ServeMux, string, []grpc.DialOption) error
+	gwfn, ok := gwRegFn.(gatewayRegisterFunctionType)
+	if !ok {
+		// Create a dummyFn only so we can ensure the correct type is shown in case
+		// of an error.
+		var dummyFn gatewayRegisterFunctionType = func(context.Context, *runtime.ServeMux, string, []grpc.DialOption) error { return nil }
+		return fmt.Errorf("unable to use %q in plugin %q due to mismatched signature.\nwant: %T\ngot: %T", gatewayRegisterFunction, pluginPath, dummyFn, gwRegFn)
+	}
+	return gwfn(gwArgs.ctx, gwArgs.mux, gwArgs.addr, gwArgs.dialOptions)
+}
+
+// listSOFiles returns the absolute paths of all .so files found in any of the provided plugin directories.
+//
+// pluginDirs can be relative to the current directory or absolute.
+func listSOFiles(fsys fs.FS, pluginDirs []string) ([]string, error) {
```

Comment:
I guess during these early dev stages no, but do you foresee any security concern loading arbitrary so files? Should we allowlisting the plugins in the values.yaml instead?

Reply:
I don't think so, in that they won't load and we'll error, unless they have the required functions for the plugin, in which case, it is a plugin in the directory specified by the caller, whether it's evil or not. Not sure, need to think more.